docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,333 @@
+since `pullback_at` only has the 2 arguments,
+and conceptually the `PullbackMemory` is a more fixed thing -- it is fully determined by the end of the primal pass,
+then during the reverse pass the argument `ȳ` is successively computed.
+We can do that via call overloading:
+```julia
+y = f(x)  # primal program
+y, pb = augmented_primal(f, x)
+x̄ = pb(ȳ)
+```
+and for `sin`:
+```julia
+y = sin(x)
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, PullbackMemory(sin; cx=cx)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * pb.cx
+```
+
+Those looking closely will spot what we have done here.
+We now have a object (`pb`) that acts on the cotangent of the output of the primal (`ȳ`) to give us the cotangent of the input of the primal function (`x̄`).
+_`pb` is not just the **memory** of state required for the `pullback`, it **is** the pullback._
+
+We have one final thing to do.
+Lets think about make the code easy to modify.
+Lets go back and think about the changes we would have to make to swap to using `sincos` from our original `sin` then `cos` way of writing.
+To rewrite that original formulation in the new pullback form we have:
+```julia
+y = sin(x)
+function augmented_primal(::typeof(sin), x)
+  y = sin(x)
+  return y, PullbackMemory(sin; x=x)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * cos(pb.x)
+```
+To go from that to:
+```julia
+y = sin(x)
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, PullbackMemory(sin; cx=cx)
+end
+(pb::PullbackMemory)(ȳ) = ȳ * pb.cx
+```
+we need to make a series of changes.
+We need to update what work is done in the primal to compute `cx`.
+We need to update what was stored in the `PullbackMemory`.
+And we need to update the the function that applies the pullback so it uses the new thing that was stored.
+It's important these thing all stay in sync.
+It's not to bad for this simple example with just 1 thing to remember.
+For more complicated multi-argument functions, like will be talked about below, you often end up needing to remember half a dozen things, like sizes and indices relating to each input/output.
+So it gets a little more fiddly to make sure you remember all the things you need to a and give them same name in both places.
+_Is there a way we can automatically just have all the things we use remembered for us?_
+
+Surprisingly for such a specific request, there actually is.
+This is a closure.
+A closure in julia is a callable structure, that automatically contains a field for every object from its parent scope that is used in its body.
+There are [incredible ways to abuse this](https://invenia.github.io/blog/2019/10/30/julialang-features-part-1#closures-give-us-classic-object-oriented-programming); but here we can in-fact use closures exactly as they are intended.
+Replacing `PullbackMemory` with a closure that works the same lets us avoid manually controlling what is remembers, _and_ lets us avoid writing separately the call overload.
```

Comment:
"is remembers" should be "it remembers" or maybe "is remembered"?

Reply:
```suggestion
Replacing `PullbackMemory` with a closure that works the same lets us avoid manually controlling what it remembers, _and_ lets us avoid writing separately the call overload.
```